flexmeasures/data/migrations/versions/cb8df44ebda5_flexcontext_field.py

```diff
@@ -369,10 +371,12 @@ def downgrade():
             else None
         )
 
-        if capacity_in_mw is not None:
-            if isinstance(capacity_in_mw, str) and is_power_unit(capacity_in_mw):
-                capacity_in_mw = (
-                    ur.Quantity(capacity_in_mw).to(ur.Quantity("MW")).magnitude
+        if site_power_capacity is not None:
+            if isinstance(site_power_capacity, str) and is_power_unit(
+                site_power_capacity
+            ):
+                site_power_capacity = (
+                    ur.Quantity(site_power_capacity).to(ur.Quantity("MW")).magnitude
                 )
```

Comment:
This is the pattern I had in mind:
```suggestion
        if site_power_capacity is not None:
            if isinstance(site_power_capacity, str) and is_power_unit(
                site_power_capacity
            ):
                site_power_capacity = (
                    ur.Quantity(site_power_capacity).to(ur.Quantity("MW")).magnitude
                )
                attributes_data["capacity_in_mw"] = capacity_in_mw
            else:
                attributes_data["site_power_capacity"] = site_power_capacity
```
The general idea is:
- If the value of the new db-flex-context field can be converted back to the expected type of the old attributes field and remain functional, we do that. This way, hosts won't lose the associated functionality when downgrading.
- Else, we just store the new field as an attribute. It won't be functionally used for anything other than the upgrade. We do this to not lose data when hosts need to downgrade and upgrade again.

Reply:
Likewise for the other fields.

Reply:
the other fields? For the other fields they already have only one consistent key unlike the capacity_in_w/site_power_capacity, if their values can't be converted on downgrade then the values are placed as-is

Reply:
> if their values can't be converted on downgrade then the values are placed as-is

Yes, so those then have the same problem that we just solved with the site-power-capacity field. If the downgraded value does not fit the downgraded field, it would break functionality. Therefore, we should store such unfit values on a separate field. My suggestion for the separate field names would be to use the serialized field names we also use in the flex context.

For example, when downgrading the `site-consumption-capacity` field:
- If the field describes a fixed power quantity, we convert it to MW and store its magnitude on the `consumption_capacity_in_mw` attribute. This is a functional attribute name (that is, scheduling functionality depends on it).
- If the field contains anything else, we store it on the `site-consumption-capacity` attribute. This is a non-functional attribute name (that is, only used to retain the user's site configuration).

Here is a list of the functional and non-functional attribute names for each flex-context field I'd want on downgrading:

- `site-consumption-capacity`
  - Functional: `consumption_capacity_in_mw` (When? If the flex-context describes a fixed quantity.)
  - Non-functional: `site-consumption-capacity`
- `site-production-capacity`
  - Functional: `production_capacity_in_mw` (When? If the flex-context describes a fixed quantity.)
  - Non-functional: `site-production-capacity`
- `site-power-capacity`
  - Functional: `capacity_in_mw` (When? If the flex-context describes a fixed quantity.)
  - Non-functional: `site-power-capacity` (NB you used underscores for this, where I prefer dashes)
- `site-peak-consumption-price`
  - Functional: `ems-peak-consumption-price` (Yes, with dashes, see [here](https://github.com/FlexMeasures/flexmeasures/pull/1293/files#diff-e35e07dd002c6a98a9e07cfbe7b7621fb762b185d04792e39f404d6e7e355f5bL295) why.) (When? If the flex-context describes a fixed quantity.)
  - Non-functional: `site-peak-consumption-price`
- `site-peak-production-price`
  - Functional: `ems-peak-production-price` (Yes, with dashes, see [here](https://github.com/FlexMeasures/flexmeasures/pull/1293/files#diff-e35e07dd002c6a98a9e07cfbe7b7621fb762b185d04792e39f404d6e7e355f5bL338) why.) (When? If the flex-context describes a fixed quantity.)
  - Non-functional: `site-peak-production-price`
- `site-consumption-breach-price`
  - Functional: `ems-consumption-breach-price` (Yes, with dashes, see [here](https://github.com/FlexMeasures/flexmeasures/pull/1293/files#diff-e35e07dd002c6a98a9e07cfbe7b7621fb762b185d04792e39f404d6e7e355f5bL383) why.) (When? If the flex-context describes a fixed quantity.)
  - Non-functional: `site-consumption-breach-price`
- `site-peak-consumption`
  -  Functional: was never functional (see [here](https://github.com/FlexMeasures/flexmeasures/pull/1293/files#diff-e35e07dd002c6a98a9e07cfbe7b7621fb762b185d04792e39f404d6e7e355f5bR261))
  - Non-functional: `site-peak-consumption`
- `site-peak-production`
  -  Functional: was never functional (see [here](https://github.com/FlexMeasures/flexmeasures/pull/1293/files#diff-e35e07dd002c6a98a9e07cfbe7b7621fb762b185d04792e39f404d6e7e355f5bR305))
  - Non-functional: `site-peak-production`

In fact, what would be wiser is to, upon downgrading, retain the entire contents of the flex-context column under an attribute named flex-context. That way we would also retain any flex-context fields added in the future. So then, upon upgrading, we would build up the new flex-context column by first taking an empty dictionary or the flex-context attribute, if it exists, and then override any flex-context fields with functional attributes.

Let me know if this is becoming too tricky, in which case I'd offer to take over your PR.

Reply:
I see, its not too complicated, ill work on this

Reply:
>This is the pattern I had in mind:

compared to your other examples, why is this "site_power_capacity" and not "site-power-capacity"